SnowplowTracker/SnowplowTracker/Storage/EventStore.cs

```diff
@@ -0,0 +1,296 @@
+		private static readonly string SELECT_ROW = "SELECT * FROM " + TABLE_NAME + " WHERE " + COLUMN_ID + "=";
+		private static readonly string DELETE_ALL = "DELETE FROM " + TABLE_NAME;
+		private static readonly string DELETE_ROW = "DELETE FROM " + TABLE_NAME + " WHERE " + COLUMN_ID + "=";
+
+		// Connection
+		private IDbConnection dbConnection;
+
+		// Add Queue
+		private Thread payloadQueueConsumer;
+		private volatile bool payloadQueueConsumerContinue = true;
+		private ConcurrentQueue<TrackerPayload> payloadQueue = new ConcurrentQueue<TrackerPayload>();
+
+		/// <summary>
+		/// Initializes a new instance of the <see cref="SnowplowTracker.Storage.EventStore"/> class.
+		/// </summary>
+		public EventStore() {
+			Open ();
+			payloadQueueConsumer = new Thread(new ThreadStart(PayloadQueueConsumer));
+			payloadQueueConsumer.Start ();
+		}
+
+		/// <summary>
+		/// Opens a connection and create's the events tables if required.
+		/// </summary>
+		public void Open() {
+			if (!IsDatabaseOpen ()) {
+				Log.Verbose("EventStore: opening database: " + DATABASE_NAME);
+				dbConnection = new SqliteConnection (DATABASE_NAME);
+				dbConnection.Open();
+
+				IDbCommand command = dbConnection.CreateCommand ();
+
+				// Set database mode to Write Ahead Logging
+				command.CommandText = "PRAGMA JOURNAL_MODE = WAL;";
+				command.ExecuteNonQuery();
+
+				// Set database mode to Serialized for safe multi-threading
+				command.CommandText = "PRAGMA SQLITE_THREADSAFE = 1;";
+				command.ExecuteNonQuery();
+
+				// Create the Table if it does not exist
+				command.CommandText = CREATE_TABLE;
+				command.ExecuteNonQuery();
+
+				Log.Verbose(" + Database open status: " + IsDatabaseOpen ());
+
+				// Close command
+				command.Dispose();
+				command = null;
+			}
+		}
+
+		/// <summary>
+		/// Closes the database connection if is currently active.
+		/// </summary>
+		public void Close() {
+			if (IsDatabaseOpen ()) {
+				dbConnection.Close();
+				dbConnection.Dispose();
+				dbConnection = null;
+			}
+
+			// Close consumers
+			payloadQueueConsumerContinue = false;
+		}
+
+		// --- Database Functions
+
+		/// <summary>
+		/// Adds the event to a concurrent queue.
+		/// </summary>
+		/// <param name="payload">Payload.</param>
+		public void AddEvent(TrackerPayload payload) {
+			payloadQueue.Enqueue (payload);
+		}
+
+		/// <summary>
+		/// Adds an event payload
+		/// </summary>
+		/// <returns><c>true</c>, if event was added, <c>false</c> otherwise.</returns>
+		/// <param name="payload">An event payload</param>
+		private void PayloadQueueConsumer() {
+			TrackerPayload payload = payloadQueue.Dequeue ();
+			bool result = false;
+			if (IsDatabaseOpen ()) {
+				byte[] bytes = Utils.SerializeDictionary (payload.GetDictionary());
+				if (bytes != null) {
+					// Build Insert Command
+					IDbCommand command = dbConnection.CreateCommand ();
+					command.CommandText = INSERT_EVENT + "(@payload)";
```

Comment:
[This post](http://zetcode.com/db/sqlitecsharp/intro/) talks about "prepared statements" increasing performance. Do you think that would apply here?